### PR TITLE
doc (typescript.md): Add missing trans in English for TypeScript

### PR DESCRIPTION
**Pull request by Maledong**
1) Add the missing English translations.
2) Format the Chinese TypeScript to make better look.

- [x] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows commit guidelines

### CODE REVIEW

**Comment by whxaxes** [comment]
awesome!!

File: docs/source/en/tutorials/typescript.md [hunk 1]
```diff
@@ -1,3 +1,542 @@
...
+
+* **Developing principles for the application layer.**
+* **How do we let developers have no sense about these problems with the support of our tools, and keep in consistency.**
+
+For more about this tossing process, please see [[RFC] TypeScript tool support](https://github.com/eggjs/egg/issues/2272)
+
+---
+
+## Quick Start
+Be a quick initialization through the skeleton:
```
Comment (whxaxes) [inline docs/source/en/tutorials/typescript.md]: `skeleton` 改成 `boilerplate` 更合理一些吧？

Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: OK. I didn't think of that :) Thanks!

File: docs/source/en/tutorials/typescript.md [hunk 2]
```diff
@@ -1,3 +1,542 @@
...
+
+![tegg.gif](https://user-images.githubusercontent.com/227713/38358019-bf7890fa-38f6-11e8-8955-ea072ac6dc8c.gif)
+
+---
+
+## Principles of catelogs
+
+**Some constraints:**
+
+* We've no plans for re-writing Egg in TS yet.
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: - we have 不能缩写为 we've 吧
- in -> with 语义会不会更好一点

File: docs/source/en/tutorials/typescript.md [hunk 3]
```diff
@@ -1,3 +1,542 @@
...
+* Egg itself, with its plug-in, will have corresponding `index.d.ts` for users to use easily.
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: its plug-in -> its plugins

File: docs/source/en/tutorials/typescript.md [hunk 4]
```diff
@@ -1,3 +1,542 @@
...
+* TypeScript only belongs to a communication practise. We support it to some extent with our tool chain.
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: practise -> practice

File: docs/source/en/tutorials/typescript.md [hunk 5]
```diff
@@ -1,3 +1,542 @@
...
+* TypeScript's version MUST BE 2.8 at least.
+
+There's no obvious difference between TS's project and Egg's common js one:
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: Egg's common js one 感觉很怪

File: docs/source/en/tutorials/typescript.md [hunk 6]
```diff
@@ -1,3 +1,542 @@
...
+import { Context } from 'egg';
+
+export default function robotMiddleware() {
+  return async (ctx: Context, next: any) => {
+    await next();
+  };
+}
+```
+
+Middlewares support input parameters, and the first one is the config of the same name. If you have other requirements, a full version is needed:
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: Middleware 和 config 都需要高亮一下

File: docs/source/en/tutorials/typescript.md [hunk 7]
```diff
@@ -1,3 +1,542 @@
...
+* In `config.{env}.ts`, we can use customized configuration settings with intellisenses in `config.default.ts`.
+
+```typescript
+// app/config/config.default.ts
+import { EggAppInfo, EggAppConfig, PowerPartial } from 'egg';
+
+// For config.{env}.ts
+export type DefaultConfig = PowerPartial<EggAppConfig & BizConfig>;
+
+// Config of App itself's Scheme
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: itself's 就有本身的意思，为啥还要加一个 's

File: docs/source/en/tutorials/typescript.md [hunk 8]
```diff
@@ -1,3 +1,542 @@
...
+  news: {
+    pageSize: number;
+    serverUrl: string;
+  };
+}
+
+export default (appInfo: EggAppInfo) => {
+  const config = {} as PowerPartial<EggAppConfig> & BizConfig;
+
+  // Override the framework, plug-in's configurations
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: plug-in 同上

File: docs/source/en/tutorials/typescript.md [hunk 9]
```diff
@@ -1,3 +1,542 @@
...
+
+## Developing period
+
+### ts-node
+
+`egg-bin` has built [ts-node](https://github.com/TypeStrong/ts-node) in, and `egg loader` will automatically load `*.ts` and compile them in memory during the period of development.
+
+Now `dev` / `debug` / `test` / `cov` are supported.
+
+For a developer's simple config in `package.json`:
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: 这个好怪， 直接 developers only need to config in .... simply

File: docs/source/en/tutorials/typescript.md [hunk 10]
```diff
@@ -1,3 +1,542 @@
...
+  "name": "showcase",
+  "egg": {
+    "typescript": true
+  }
+}
+```
+
+### egg-ts-helper
+
+Due to the automatic loading mechanisim, TS cannot analyze dependencies in static or relationship intellisenses.
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: mechanisim -> mechanism

File: docs/source/en/tutorials/typescript.md [hunk 11]
```diff
@@ -1,3 +1,542 @@
...
+import News from '../../../app/service/News';
+
+declare module 'egg' {
+  interface IService {
+    news: News;
+  }
+}
+```
+
+It's a bit too bothering to write them manually, so we offer you a tool [egg-ts-helper](https://github.com/whxaxes/egg-ts-helper), with which we can analyze and generate `d.ts` files automatically.
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: a bit , too 好中文化。。。。

File: docs/source/en/tutorials/typescript.md [hunk 12]
```diff
@@ -1,3 +1,542 @@
...
+  },
+  "exclude": [
+    "app/public",
+    "app/web",
+    "app/views"
+  ]
+}
+```
+
+**Notice: When ts and js files of the same names, egg will first load js ones. So during the development period, `egg-ts-helper` will be automatically called to clear up all the js files with the same names, of course you can use `npm run clean` to clear them manually.**
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: When there are ts and js files with the same name

File: docs/source/en/tutorials/typescript.md [hunk 13]
```diff
@@ -1,3 +1,542 @@
...
+
+### Error Stacks
+
+Codes online are js after compliation, however what we expect is to see error stacks pointing at TS source codes. So:
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: - compliation -> compilation
- pointing 为啥要用进行时

File: docs/source/en/tutorials/typescript.md [hunk 14]
```diff
@@ -1,3 +1,542 @@
...
+
+* When buiding the project, we should insert info of `sourcemap` in `inlineSourceMap: true`.
+* There's no worry for developers to worry about correcting the positions to the right error stacks in a built-in `egg-scripts`.
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: no worry, worry about 重复表述

File: docs/source/en/tutorials/typescript.md [hunk 15]
```diff
@@ -1,3 +1,542 @@
...
+
+For more detailed info:
+
+* [https://zhuanlan.zhihu.com/p/26267678](https://zhuanlan.zhihu.com/p/26267678)
+* [https://github.com/eggjs/egg-scripts/pull/19](https://github.com/eggjs/egg-scripts/pull/19)
+
+---
+
+## Guides to the developments of Plug-in/Framework
+
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: to 换成 more 是不是好一点

File: docs/source/en/tutorials/typescript.md [hunk 16]
```diff
@@ -1,3 +1,542 @@
...
+**Principles:**
+
+* DO NOT recommand to develop plug-in/framework in TS directly, we should publish them in js.
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: - recommand -> recommend
- plug-in -> plugin
- 发布到 npm 好像没翻译

File: docs/source/en/tutorials/typescript.md [hunk 17]
```diff
@@ -1,3 +1,542 @@
...
+* When you write a plug-in/framwork, the corresponding `index.d.ts` should be included.
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: framwork -> framework

File: docs/source/en/tutorials/typescript.md [hunk 18]
```diff
@@ -1,3 +1,542 @@
...
+* By [Declaration Merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) we can inject the functions of plug-in/framework into Egg.
+* All are mounted on `egg` module, DO NOT use the outer layer.
+
+### Plug-in
+
+Styles can referred from the automatically generated `egg-ts-helper`:
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: referred -> refer

File: docs/source/en/tutorials/typescript.md [hunk 19]
```diff
@@ -1,3 +1,542 @@
...
+  }
+
+  // extend customize env
+  type EggEnvType = 'local' | 'unittest' | 'prod' | 'sit';
+}
+```
+
+### The outer framework
+
+Definations:
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: definations -> definitions

File: docs/source/en/tutorials/typescript.md [hunk 20]
```diff
@@ -1,3 +1,542 @@
...
+// Export the whole Egg
+export = Egg;
+```
+
+For developers, they can directly import your framework:
+
+```typescript
+// app/service/news.ts
+
+// 开发者引入你的框架，也可以使用到提示到所有 Egg 的提示
```
Comment (SunShinewyf) [inline docs/source/en/tutorials/typescript.md]: 这个没翻译

File: docs/source/en/tutorials/typescript.md [hunk 21]
```diff
@@ -1,3 +1,542 @@
...
-this document is still waiting for translation, see [Chinese Version](/zh-cn/tutorials/typescript.html)
\ No newline at end of file
+
+> [TypeScript](https://www.typescriptlang.org/) is a typed superset of JavaScript that compiles to plain JavaScript.
+
+For a large number of enterprises' applications, TypeScript's static type checking, intellisenses, friendly IDE are valuable. For more please see [System Research Report For TypeScript](https://juejin.im/post/59c46bc86fb9a00a4636f939) 。
+
+However, we've met some problems influencing users' experience when developing Egg in TypeScript: 
+
+* The most outstanding Loader Mechanism (Auto-loading) makes TS not analyze dependencies in static.
```
[the remark this answers is not in the document]
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: Grammar error：Make sb/sth not do sth

File: docs/source/en/tutorials/typescript.md [hunk 22]
```diff
@@ -1,3 +1,542 @@
-title: Assets
+title: TypeScript
 ---
-this document is still waiting for translation, see [Chinese Version](/zh-cn/tutorials/typescript.html)
\ No newline at end of file
+
+> [TypeScript](https://www.typescriptlang.org/) is a typed superset of JavaScript that compiles to plain JavaScript.
+
+For a large number of enterprises' applications, TypeScript's static type checking, intellisenses, friendly IDE are valuable. For more please see [System Research Report For TypeScript](https://juejin.im/post/59c46bc86fb9a00a4636f939) 。
```
[the remark this answers is not in the document]
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: 智能提示确实是“intellisense”

File: docs/source/en/tutorials/typescript.md [hunk 23]
```diff
@@ -1,3 +1,542 @@
...
+
+However, we've met some problems influencing users' experience when developing Egg in TypeScript: 
+
+* The most outstanding Loader Mechanism (Auto-loading) makes TS not analyze dependencies in static.
+* How to validate and show intellisenses in `config.{env}.js`, when we modify settings by plug-in and these configurations are automatically merged? 
+* During the period of developing, `tsc -w` is created as an independent process to build up codes, it makes us of two minds about where to save the temporary files, and the complicated `npm scripts`.
```
[the remark this answers is not in the document]
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: be of two minds……本身就是“纠结于……”的含义。请注意此用法：https://www.oxfordlearnersdictionaries.com/definition/english/mind_1#mind_1__366，你可以搜索一下。不过就按照你的为准，如果这种用法不多见的话。

File: docs/source/en/tutorials/typescript.md [hunk 24]
```diff
@@ -1,3 +1,542 @@
...
+## Quick Start
+Be a quick initialization through the skeleton:
+
+```bash
+$ npx egg-init --type=ts showcase
+$ cd showcase && npm i
+$ npm run dev
+```
+
+The skeleton above can be a very simple example, for a detailed one please see[eggjs/examples/hackernews-async-ts](https://github.com/eggjs/examples/tree/master/hackernews-async-ts)
```
[the remark this answers is not in the document]
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: Fixed.

File: docs/source/en/tutorials/typescript.md [hunk 25]
```diff
@@ -1,3 +1,542 @@
...
+
+![tegg.gif](https://user-images.githubusercontent.com/227713/38358019-bf7890fa-38f6-11e8-8955-ea072ac6dc8c.gif)
+
+---
+
+## Principles of catelogs
```
[the remark this answers is not in the document]
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: Fixed, thanks!

File: docs/source/en/tutorials/typescript.md (hunk 2 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: 1) We have = We've，是可以的。
2) in 某种语言。E.g: in English……TS是一种语言，而with一般是用于某种工具。

File: docs/source/en/tutorials/typescript.md (hunk 3 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: 其余英语翻译部分都是plug-in，这里保持一致就不改了。

File: docs/source/en/tutorials/typescript.md (hunk 4 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: Fixed.

File: docs/source/en/tutorials/typescript.md (hunk 5 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: There's no obvious difference between TS's project and Egg's in js

File: docs/source/en/tutorials/typescript.md (hunk 6 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: OK

File: docs/source/en/tutorials/typescript.md (hunk 7 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: Config of App's Scheme

File: docs/source/en/tutorials/typescript.md (hunk 8 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: 保持一致，就不改了。

File: docs/source/en/tutorials/typescript.md (hunk 9 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: OK

File: docs/source/en/tutorials/typescript.md (hunk 10 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: Sorry and fixed.

File: docs/source/en/tutorials/typescript.md (hunk 12 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: there be的句型不是单独这样使用的。不能直接说 There's a book。因为“存在句”只有一个主语（a book），所以必须加上状语（如 There's a book there）等。

所以这里改成：When ts and js files are with the same names……

File: docs/source/en/tutorials/typescript.md (hunk 11 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: 的确英语中也有此用法：I feel a bit too tired（而且 a bit too略带贬义的含义）。

File: docs/source/en/tutorials/typescript.md (hunk 13 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: see sth/sb doing sth，很正常的语法。也符合语境。

File: docs/source/en/tutorials/typescript.md (hunk 14 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: For developers, there's no need to worry about correcting the positions to the right error stacks in a built-in `egg-scripts`.

File: docs/source/en/tutorials/typescript.md (hunk 15 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: Guides more……？不通顺的，没有介词连接。这个中文的意思是：插件/框架开发指南。to在这里表示针对……而言。比较恰当呢，谢谢。

File: docs/source/en/tutorials/typescript.md (hunk 16 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: OK.

File: docs/source/en/tutorials/typescript.md (hunk 17 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: OK

File: docs/source/en/tutorials/typescript.md (hunk 18 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: be referred.. Fixed. Thanks!

File: docs/source/en/tutorials/typescript.md (hunk 19 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: OK

File: docs/source/en/tutorials/typescript.md (hunk 20 above)
Comment (Maledong) [reply docs/source/en/tutorials/typescript.md]: OK. Fixed.

**Comment by Maledong** [comment]
@SunShinewyf：Thanks for you GREAT contributions of review. I've made some comments for you and do some fixtures.

File: docs/source/en/tutorials/typescript.md (hunk 15 above)
Comment (SunShinewyf) [reply docs/source/en/tutorials/typescript.md]: 说错了，to->for

File: docs/source/en/tutorials/typescript.md [hunk 26]
```diff
@@ -1,3 +1,542 @@
...
\ No newline at end of file
+
+> [TypeScript](https://www.typescriptlang.org/) is a typed superset of JavaScript that compiles to plain JavaScript.
+
+For a large number of enterprises' applications, TypeScript's static type checking, intellisenses, friendly IDE are valuable. For more please see [System Research Report For TypeScript](https://juejin.im/post/59c46bc86fb9a00a4636f939) 。
+
+However, we've met some problems influencing users' experience when developing Egg in TypeScript: 
+
+* The most outstanding Loader Mechanism (Auto-loading) makes TS not analyze dependencies in static.
+* How to validate and show intellisenses in `config.{env}.js`, when we modify settings by plug-in and these configurations are automatically merged?
```
[the remark this answers is not in the document]
Comment (SunShinewyf) [reply docs/source/en/tutorials/typescript.md]: ok

File: docs/source/en/tutorials/typescript.md (hunk 12 above)
Comment (SunShinewyf) [reply docs/source/en/tutorials/typescript.md]: ok

**Comment by atian25** [comment]
@Maledong rebase

@SunShinewyf is it ready to  merge?

**Comment by SunShinewyf** [comment]
OK

**Comment by SunShinewyf** [comment]
OK

**Comment by Maledong** [comment]
@SunShinewyf：Thanks anyway!

**Comment by Maledong** [comment]
Thanks all

**Pull request merged**